quantum/matrix.c

```diff
@@ -102,7 +102,7 @@ __attribute__((weak)) void matrix_init_pins(void) {
     for (int row = 0; row < MATRIX_ROWS; row++) {
         for (int col = 0; col < MATRIX_COLS; col++) {
             pin_t pin = direct_pins[row][col];
-            if (pin != NO_PIN) {
+            if ((pin != NO_PIN) && (pin != 0)) {
```

Comment:
While i was testing with a.dux keyboard, i found VIA not able to save changed keymaps.

Can someone please check whether configuration in this keyboard is correct ?
https://github.com/qmk/qmk_firmware/blob/master/keyboards/a_dux/config.h#L29
https://github.com/qmk/qmk_firmware/blob/master/keyboards/a_dux/config.h#L43

The issue is number of rows mismatch in `MATRIX_ROWS` and `DIRECT_PINS`. 
matrix.c defines an array of size MATRIX_ROWS which is 8 and direct_pins only define four rows. So the remaining array elements are initialized to 0.
Later on, when the pins are inited here https://github.com/qmk/qmk_firmware/blob/master/quantum/matrix.c#L113 
setPinInputHigh(0) is called. This might not be a problem on AVR chips. But makes STM chip go nuts. It sets the FLASH->SR  status to 0xC0 and not able to reset manually too.
This prevents the further flash write operation to fail.

I quickly changed only the place which i needed to proceed further. There are other places too, where we are initing pins, which could benefit from this check.

I am not sure though whether adux config.h is correct. So did not bother to create a new PR.
If someone can confirm this, i can create a new PR with proper explanation and fix.

Related links:
https://community.st.com/s/question/0D50X00009XkaGESAZ/stm2f4xx-st3240g-cant-clear-flash-error-flags

Reply:
> The issue is number of rows mismatch in MATRIX_ROWS and DIRECT_PINS.

This is mostly expected from a configuration stance, as can been seen in various keyboards which use `split common` (for example <https://github.com/qmk/qmk_firmware/blob/master/keyboards/lets_split/rev2/config.h#L29>)

<https://github.com/qmk/qmk_firmware/pull/16292> should hopefully resolve the issue.

Reply:
I think its a valid assumption. Thanks.
Is there any reason/advantage to specify NO_PIN as `~0` instead of `0` ?